modules/ballerina-core/src/main/java/org/ballerinalang/model/util/JsonNode.java

```diff
@@ -0,0 +1,398 @@
+        this.booleanValue = booleanValue;
+        this.type = Type.BOOLEAN;
+    }
+    
+    public void setNull() {
+        this.type = Type.NULL;
+    }
+    
+    public boolean isNull() {
+        return this.type == Type.NULL;
+    }
+    
+    public boolean isObject() {
+        return this.type == Type.OBJECT;
+    }
+    
+    public boolean isArray() {
+        return this.type == Type.ARRAY;
+    }
+    
+    public boolean isValueNode() {
+        return this.isString() || this.isLong() || this.isDouble() || this.isBoolean() || this.isNull();
+    }
+    
+    public boolean isLong() {
+        return this.type == Type.LONG;
+    }
+    
+    public boolean isDouble() {
+        return this.type == Type.DOUBLE;
+    }
+    
+    public boolean isString() {
+        return this.type == Type.STRING;
+    }
+    
+    public boolean isBoolean() {
+        return this.type == Type.BOOLEAN;
+    }
+    
+    public int size() {
+        if (this.arrayElements != null) {
+            return this.arrayElements.size();
+        } else {
+            return 0;
+        }
+    }
+    
+    public String stringValue() {
+        return stringValue;
+    }
+    
+    public long longValue() {
+        return longValue;
+    }
+    
+    public double doubleValue() {
+        return doubleValue;
+    }
+    
+    public boolean booleanValue() {
+        return booleanValue;
+    }
+    
+    public void set(String name, String stringValue) {
+        if (this.fields == null) {
+            this.fields = new LinkedHashMap<>();
+        }
+        if (stringValue == null) {
+            this.fields.put(name, new JsonNode(Type.NULL));
+        } else {
+            this.fields.put(name, new JsonNode(stringValue));
+        }
+    }
+    
+    public void set(String name, long longValue) {
+        if (this.fields == null) {
+            this.fields = new LinkedHashMap<>();
+        }
+        this.fields.put(name, new JsonNode(longValue));
+    }
+    
+    public void set(String name, double doubleValue) {
+        if (this.fields == null) {
+            this.fields = new LinkedHashMap<>();
+        }
+        this.fields.put(name, new JsonNode(doubleValue));
+    }
+    
+    public void set(String name, boolean booleanValue) {
+        if (this.fields == null) {
+            this.fields = new LinkedHashMap<>();
+        }
+        this.fields.put(name, new JsonNode(booleanValue));
+    }
+    
+    public void set(String name, JsonNode jsonNode) {
+        if (this.fields == null) {
+            this.fields = new LinkedHashMap<>();
+        }
+        if (jsonNode != null) {
+            this.fields.put(name, jsonNode);
+        } else {
+            this.fields.put(name, new JsonNode(Type.NULL));
+        }
+    }
+    
+    public void set(int index, JsonNode jsonNode) {
+        if (this.arrayElements == null) {
+            throw new BallerinaException("JSON array is empty to set values");
```

Comment:
Is this error message correct? Is the JSON array is empty to null?

Reply:
the "arrayElements" is null initially as an optimization technique, where it is only initialized if needed. So when a JsonNode is created as an ARRAY type, it has zero elements, but the "arrayElements" can be still null. It will be initialized when an actual element is put into the object. This is because, a JsonNode represents all types, object, value types, arrays etc.. so initializing variables to support all of these at a single time is not needed.

Reply:
ACK